rust/src/http2/detect.rs

```diff
@@ -0,0 +1,399 @@
+            }
+            _ => {
+                return 0;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let value = &hd.blocks[i as usize].name;
+                    unsafe {
+                        *buffer = value.as_ptr();
+                        *buffer_len = value.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+            _ => {
+                return 0;
+            }
+        },
+        _ => {}
+    }
+
+    return 0;
+}
+
+fn http2_escape_header(hd: &parser::HTTP2FrameHeaders, i: u32) -> Vec<u8> {
+    //minimum size + 2 for escapes
+    let normalsize = hd.blocks[i as usize].value.len() + 2 + hd.blocks[i as usize].name.len() + 2;
+    let mut vec = Vec::with_capacity(normalsize);
+    for j in 0..hd.blocks[i as usize].name.len() {
+        vec.push(hd.blocks[i as usize].name[j]);
+        if hd.blocks[i as usize].name[j] == ':' as u8 {
+            vec.push(':' as u8);
+        }
+    }
+    vec.push(':' as u8);
+    vec.push(' ' as u8);
+    for j in 0..hd.blocks[i as usize].value.len() {
+        vec.push(hd.blocks[i as usize].value[j]);
+        if hd.blocks[i as usize].value[j] == ':' as u8 {
+            vec.push(':' as u8);
+        }
+    }
+    return vec;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_get_header(
+    tx: *mut std::os::raw::c_void, direction: u8, i: u32, buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    match direction {
+        STREAM_TOSERVER => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let vec = http2_escape_header(hd, i);
+                    let value = &vec;
+                    unsafe {
+                        *buffer = value.as_ptr();
+                        *buffer_len = value.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+            _ => {
+                return 0;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let mut vec = Vec::new();
+                    for j in 0..hd.blocks[i as usize].name.len() {
+                        vec.push(hd.blocks[i as usize].name[j]);
+                        /* this escaping is needed
+                         * to avoid amibguity with separator between name and value
+                         */
+                        if hd.blocks[i as usize].name[j] == ':' as u8 {
+                            vec.push(':' as u8);
+                        }
+                    }
+                    vec.push(':' as u8);
+                    vec.push(' ' as u8);
```

Comment:
very minor, but I wonder if this would save a few cycles:
```
    vec.extend(&[':' as u8, ' ' as u8]);
```

Reply:
So, leaving this way for now